File: packages/core/src/patterns/KajayPatternParser.ts
import {
  alternation,
  EMPTY_PATTERN,
  sequence,
  type CharacterClassItem,
  type PatternNode,
  type ScalarPattern,
  type ShorthandName,
} from './KajayPatternNode.js';

const MAX_SOURCE_SCALARS = 512;
const MAX_REPETITION = 1_000;
const SHORTHAND_ESCAPES = new Set<string>(['d', 'D', 'w', 'W', 's', 'S']);

export function parseKajayPattern(source: string): PatternNode | undefined {
  let count = 0;
  for (const scalar of source) {
    const value = scalar.codePointAt(0)!;
    count += 1;
    if (count > MAX_SOURCE_SCALARS || (value >= 0xd800 && value <= 0xdfff)) {
      return undefined;
    }
  }
  return new PatternParser(source).parse();
}

class PatternParser {
  readonly #source: string;
  #index = 0;
  #valid = true;

  constructor(source: string) {
    this.#source = source;
  }

  parse(): PatternNode | undefined {
    const node = this.#alternation();
    return this.#valid && this.#index === this.#source.length ? node : undefined;
  }

  #alternation(terminator?: string): PatternNode {
    const alternatives = [this.#sequence(terminator)];
    while (this.#peek() === '|') {
      this.#index += 1;
      alternatives.push(this.#sequence(terminator));
    }
    return alternation(alternatives);
  }

  #sequence(terminator?: string): PatternNode {
    const items: PatternNode[] = [];
    while (this.#index < this.#source.length) {
      const current = this.#peek();
      if (current === '|' || current === terminator) {
        break;
      }
      items.push(this.#atom());
    }
    return sequence(items);
  }

  #atom(): PatternNode {
    const current = this.#take();
    let node: PatternNode;
    switch (current) {
      case undefined:
      case ')':
      case ']':
      case '}':
      case '*':
      case '+':
      case '?':
      case '{':
        return this.#fail();
      case '^':
        return { kind: 'anchor', edge: 'start' };
      case '$':
        return { kind: 'anchor', edge: 'end' };
      case '(':
        if (this.#peek() === '?') {
          return this.#fail();
        }
        node = this.#alternation(')');
        if (this.#take() !== ')') {
          return this.#fail();
        }
        break;
      case '[':
        node = { kind: 'scalar', scalar: this.#characterClass() };
        break;
      case '\\': {
        const escaped = this.#escape();
        node = escaped === undefined ? this.#fail() : { kind: 'scalar', scalar: escaped };
        break;
      }
      case '.':
        node = { kind: 'scalar', scalar: { kind: 'dot' } };
        break;
      default:
        node = { kind: 'scalar', scalar: { kind: 'literal', value: current.codePointAt(0)! } };
        break;
    }
    return this.#quantifier(node);
  }

  #characterClass(): ScalarPattern {
    const negated = this.#peek() === '^';
    if (negated) {
      this.#index += 1;
    }
    const items: CharacterClassItem[] = [];
    while (this.#index < this.#source.length && this.#peek() !== ']') {
      const start = this.#classItem();
      if (start === undefined) {
        return this.#failedClass();
      }
      if (this.#peek() !== '-') {
        items.push(start);
        continue;
      }
      this.#index += 1;
      const end = this.#classItem();
      if (
        start.kind !== 'range' ||
        start.first !== start.last ||
        end?.kind !== 'range' ||
        end.first !== end.last ||
        end.first < start.first
      ) {
        return this.#failedClass();
      }
      items.push({ kind: 'range', first: start.first, last: end.first });
    }
    if (items.length === 0 || this.#take() !== ']') {
      return this.#failedClass();
    }
    return { kind: 'class', negated, items };
  }

  #classItem(): CharacterClassItem | undefined {
    const current = this.#take();
    if (current === undefined || current === ']' || current === '-') {
      return undefined;
    }
    if (current !== '\\') {
      const value = current.codePointAt(0)!;
      return { kind: 'range', first: value, last: value };
    }
    const escaped = this.#escape();
    if (escaped?.kind === 'shorthand') {
      return escaped;
    }
    return escaped?.kind === 'literal'
      ? { kind: 'range', first: escaped.value, last: escaped.value }
      : undefined;
  }

  #escape(): ScalarPattern | undefined {
    const escaped = this.#take();
    if (escaped === undefined) {
      return undefined;
    }
    if (SHORTHAND_ESCAPES.has(escaped)) {
      return { kind: 'shorthand', name: escaped as ShorthandName };
    }
    if (/^[A-Za-z0-9]$/u.test(escaped)) {
      return undefined;
    }
    return { kind: 'literal', value: escaped.codePointAt(0)! };
  }

  #quantifier(item: PatternNode): PatternNode {
    const current = this.#peek();
    if (current === '*' || current === '+' || current === '?') {
      this.#index += 1;
      return {
        kind: 'repeat',
        item,
        minimum: current === '+' ? 1 : 0,
        ...(current === '*' || current === '+' ? {} : { maximum: 1 }),
      };
    }
    if (current !== '{') {
      return item;
    }
    this.#index += 1;
    const minimum = this.#decimal();
    if (minimum === undefined) {
      return this.#fail();
    }
    let maximum = minimum;
    if (this.#peek() === ',') {
      this.#index += 1;
      maximum = this.#decimal() ?? -1;
    }
    if (this.#take() !== '}' || maximum < minimum || maximum > MAX_REPETITION) {
      return this.#fail();
    }
    return { kind: 'repeat', item, minimum, maximum };
  }

  #decimal(): number | undefined {
    const start = this.#index;
    while (/^[0-9]$/u.test(this.#peek() ?? '')) {
      this.#index += 1;
    }
    return this.#index === start ? undefined : Number(this.#source.slice(start, this.#index));
  }

  #take(): string | undefined {
    const value = this.#peek();
    if (value !== undefined) {
      this.#index += value.length;
    }
    return value;
  }

  #peek(): string | undefined {
    const point = this.#source.codePointAt(this.#index);
    return point === undefined ? undefined : String.fromCodePoint(point);
  }

  #failedClass(): ScalarPattern {
    this.#valid = false;
    return { kind: 'literal', value: 0 };
  }

  #fail(): PatternNode {
    this.#valid = false;
    return EMPTY_PATTERN;
  }
}

File: dotnet/src/Kajay.Core/Expressions/Patterns/KajayPatternParser.cs
using System.Text;

namespace Kajay.Expressions.Patterns;

internal sealed partial class KajayPatternParser
{
    private const int MaxSourceScalars = 512;
    private const int MaxRepetition = 1_000;
    private readonly string _source;
    private int _index;
    private bool _valid = true;

    private KajayPatternParser(string source)
    {
        _source = source;
    }

    public static KajayPatternNode? Parse(string source)
    {
        ArgumentNullException.ThrowIfNull(source);
        if (source.EnumerateRunes().Take(MaxSourceScalars + 1).Count() > MaxSourceScalars)
        {
            return null;
        }

        var parser = new KajayPatternParser(source);
        KajayPatternNode node = parser.ParseAlternation(null);
        return parser._valid && parser._index == source.Length ? node : null;
    }

    private KajayPatternNode ParseAlternation(char? terminator)
    {
        var alternatives = new List<KajayPatternNode> { ParseSequence(terminator) };
        while (PeekAscii() == '|')
        {
            _index += 1;
            alternatives.Add(ParseSequence(terminator));
        }

        return alternatives.Count == 1
            ? alternatives[0]
            : new KajayPatternNode.Alternation(alternatives);
    }

    private KajayPatternNode ParseSequence(char? terminator)
    {
        var items = new List<KajayPatternNode>();
        while (_index < _source.Length)
        {
            char? current = PeekAscii();
            if (current == '|' || current == terminator)
            {
                break;
            }

            items.Add(ParseAtom());
        }

        return items.Count switch
        {
            0 => new KajayPatternNode.Empty(),
            1 => items[0],
            _ => new KajayPatternNode.Sequence(items),
        };
    }

    private KajayPatternNode ParseAtom()
    {
        char? current = PeekAscii();
        if (current is null)
        {
            int? scalar = TakeScalar();
            return scalar is null
                ? Fail()
                : ParseQuantifier(new KajayPatternNode.Scalar(
                    new KajayScalarPattern.Literal(scalar.Value)));
        }

        _index += 1;
        KajayPatternNode node;
        switch (current)
        {
            case ')' or ']' or '}' or '*' or '+' or '?' or '{':
                return Fail();
            case '^':
                return new KajayPatternNode.Anchor(true);
            case '$':
                return new KajayPatternNode.Anchor(false);
            case '(':
                node = ParseGroup();
                break;
            case '[':
                node = new KajayPatternNode.Scalar(ParseCharacterClass());
                break;
            case '\\':
                KajayScalarPattern? escaped = ParseEscape();
                node = escaped is null
                    ? Fail()
                    : new KajayPatternNode.Scalar(escaped);
                break;
            case '.':
                node = new KajayPatternNode.Scalar(new KajayScalarPattern.Dot());
                break;
            default:
                node = new KajayPatternNode.Scalar(
                    new KajayScalarPattern.Literal(current.Value));
                break;
        }

        return ParseQuantifier(node);
    }

    private KajayPatternNode ParseGroup()
    {
        if (PeekAscii() == '?')
        {
            return Fail();
        }

        KajayPatternNode node = ParseAlternation(')');
        return TakeAscii(')') ? node : Fail();
    }

    private KajayScalarPattern ParseCharacterClass()
    {
        bool negated = PeekAscii() == '^';
        if (negated)
        {
            _index += 1;
        }

        var items = new List<KajayCharacterClassItem>();
        while (_index < _source.Length && PeekAscii() != ']')
        {
            KajayCharacterClassItem? start = ParseClassItem();
            if (start is null)
            {
                return FailedClass();
            }

            if (PeekAscii() != '-')
            {
                items.Add(start);
                continue;
            }

            _index += 1;
            KajayCharacterClassItem? end = ParseClassItem();
            if (start is not KajayCharacterClassItem.Range first
                || first.First != first.Last
                || end is not KajayCharacterClassItem.Range last
                || last.First != last.Last
                || last.First < first.First)
            {
                return FailedClass();
            }

            items.Add(new KajayCharacterClassItem.Range(first.First, last.First));
        }

        if (items.Count == 0 || !TakeAscii(']'))
        {
            return FailedClass();
        }

        return new KajayScalarPattern.CharacterClass(negated, items);
    }

    private KajayCharacterClassItem? ParseClassItem()
    {
        if (PeekAscii() is ']' or '-')
        {
            return null;
        }

        if (PeekAscii() != '\\')
        {
            int? scalar = TakeScalar();
            return scalar is null
                ? null
                : new KajayCharacterClassItem.Range(scalar.Value, scalar.Value);
        }

        _index += 1;
        KajayScalarPattern? escaped = ParseEscape();
        return escaped switch
        {
            KajayScalarPattern.Shorthand shorthand =>
                new KajayCharacterClassItem.Shorthand(shorthand.Name),
            KajayScalarPattern.Literal literal =>
                new KajayCharacterClassItem.Range(literal.Value, literal.Value),
            _ => null,
        };
    }

    private KajayScalarPattern? ParseEscape()
    {
        int? escaped = TakeScalar();
        if (escaped is null)
        {
            return null;
        }

        if (escaped is 'd' or 'D' or 'w' or 'W' or 's' or 'S')
        {
            return new KajayScalarPattern.Shorthand((char)escaped.Value);
        }

        var rune = new Rune(escaped.Value);
        return Rune.IsLetterOrDigit(rune)
            ? null
            : new KajayScalarPattern.Literal(escaped.Value);
    }

    private KajayPatternNode ParseQuantifier(KajayPatternNode item)
    {
        char? current = PeekAscii();
        if (current is '*' or '+' or '?')
        {
            _index += 1;
            return new KajayPatternNode.Repeat(
                item,
                current == '+' ? 1 : 0,
                current is '*' or '+' ? null : 1);
        }

        if (current != '{')
        {
            return item;
        }

        _index += 1;
        int? minimum = ParseDecimal();
        if (minimum is null)
        {
            return Fail();
        }

        int maximum = minimum.Value;
        if (PeekAscii() == ',')
        {
            _index += 1;
            maximum = ParseDecimal() ?? -1;
        }

        if (!TakeAscii('}') || maximum < minimum || maximum > MaxRepetition)
        {
            return Fail();
        }

        return new KajayPatternNode.Repeat(item, minimum.Value, maximum);
    }

}
